src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/AvoidPotentiallyExpensiveCallWhenLoggingTests.cs

```diff
@@ -0,0 +1,3532 @@
+                class C
+                {
+                    async void M(ILogger logger, EventId eventId, Exception exception, Func<object, Exception, string> formatter, Task<string> task)
+                    {
+                        logger.Log(LogLevel.Debug, eventId, [|await task|], exception, formatter);
+                    }
+                }
+                """;
+
+            await VerifyCSharpCodeFixAsync(source, source);
+        }
+
+        [Fact]
+        public async Task BinaryOperation_ReportsDiagnostic_CS()
+        {
+            string source = """
+                using System;
+                using Microsoft.Extensions.Logging;
+
+                class C
+                {
+                    void M(ILogger logger, EventId eventId, Exception exception, Func<object, Exception, string> formatter)
+                    {
+                        logger.Log(LogLevel.Debug, eventId, [|4 + 2|], exception, formatter);
+                    }
+                }
+                """;
+
+            await VerifyCSharpCodeFixAsync(source, source);
+        }
+
+        [Fact]
+        public async Task CoalesceOperation_ReportsDiagnostic_CS()
+        {
+            string source = """
+                #nullable enable
+
+                using System;
+                using Microsoft.Extensions.Logging;
+
+                class C
+                {
+                    void M(ILogger logger, EventId eventId, Exception? exception, Func<object, Exception?, string> formatter)
+                    {
+                        logger.Log(LogLevel.Debug, eventId, [|exception ?? new Exception()|], exception, formatter);
+                    }
+                }
+                """;
+
+            await VerifyCSharpCodeFixAsync(source, source);
+        }
+
+        [Fact]
+        public async Task CollectionExpressionOperation_ReportsDiagnostic_CS()
+        {
+            string source = """
+                using System;
+                using Microsoft.Extensions.Logging;
+
+                class C
+                {
+                    void M(ILogger logger, EventId eventId, Exception exception, Func<int[], Exception, string> formatter)
+                    {
+                        logger.Log(LogLevel.Debug, eventId, [|[4, 2]|], exception, formatter);
+                    }
+                }
+                """;
+
+            await VerifyCSharpCodeFixAsync(source, source, CodeAnalysis.CSharp.LanguageVersion.CSharp12);
+        }
+
+        [Fact]
+        public async Task DefaultValueOperation_ReportsDiagnostic_CS()
+        {
+            string source = """
+                using System;
+                using Microsoft.Extensions.Logging;
+
+                class C
+                {
+                    void M(ILogger logger, EventId eventId, Exception exception, Func<int, Exception, string> formatter)
+                    {
+                        logger.Log(LogLevel.Debug, eventId, [|default|], exception, formatter);
```

Comment:
Isn't `default` free at runtime?

Reply:
> Isn't `default` free at runtime?

For reference types.

For value types, it depends on the type. But it should, in general, be "free".

Reply:
Right, I mean, there's no overhead in passing the value. It's just zeroed memory. I don't think it warrants a diagnostic.

[Here's a sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgPgAQMwAIEEYBsaBMKDCAsAFADeJKlaqG2CALCgLIAUAlBVecVb2owDEA9kJYAHCACcIYAM4ohwAFYBTAMYAXAPwBtALoLl6jbLakAviU58Uw0QBMVAMwgBXADYaWsDWwA0KI4uHl6Kqpr+KDAyKkJOLHZsbADc1iiWxOZAA=) showing `default` and `nameof` are both compile-time constant values.

```c#
private static void Foo(params object?[] objects) {}

public static void M()
{
    Foo(default(int), default(object), nameof(Foo));
}
```

...`M()` becomes...

```c#
object[] array = new object[3];
array[0] = 0; // don't set array[1] as it's already null
array[2] = "Foo";
Foo(array);
```

Reply:
For this code, value types will be boxed.

Wither being boxed or passed by value, big value types may be a problem.

I know you're not thinking of those, but they are out there.

Reply:
If the original concern here was boxing (having that info in the diagnostic message, or at least in a comment would be helpful) then it should only apply to value types. I'd like to hear @mpidash's motivation here.